src/Tools/dotnet-monitor/CollectionRules/Actions/ExecuteAction.cs

```diff
@@ -16,76 +17,137 @@
         {
-            string path = options.Path;
-            string arguments = options.Arguments;
-            bool IgnoreExitCode = options.IgnoreExitCode.GetValueOrDefault(ExecuteOptionsDefaults.IgnoreExitCode);
-
-            ValidateFilePath(path);
+            if (null == options)
+            {
+                throw new ArgumentNullException(nameof(options));
+            }
 
-            // May want to capture stdout and stderr and return as part of the result in the future
-            using Process process = new Process();
+            ValidationContext context = new(options, null, items: null);
+            Validator.ValidateObject(options, context, validateAllProperties: true);
 
-            process.StartInfo = new ProcessStartInfo(path, arguments);
-            process.StartInfo.RedirectStandardOutput = true;
+            return new ExecuteAction(options);
+        }
 
-            // Completion source that is signaled when the process exits
-            TaskCompletionSource<int> exitedSource = new TaskCompletionSource<int>(TaskCreationOptions.RunContinuationsAsynchronously);
+        internal sealed class ExecuteAction :
+            ICollectionRuleAction,
+            IAsyncDisposable
+        {
+            private readonly CancellationTokenSource _disposalTokenSource = new();
+            private readonly ExecuteOptions _options;
 
-            EventHandler exitedHandler = (s, e) => exitedSource.TrySetResult(process.ExitCode);
+            private Task<CollectionRuleActionResult> _completionTask;
 
-            process.EnableRaisingEvents = true;
-            process.Exited += exitedHandler;
+            public ExecuteAction(ExecuteOptions options)
+            {
+                _options = options ?? throw new ArgumentNullException(nameof(options));
+            }
 
-            try
+            public async ValueTask DisposeAsync()
             {
-                if (!process.Start())
-                {
-                    throw new CollectionRuleActionException(new InvalidOperationException(string.Format(CultureInfo.InvariantCulture, Strings.ErrorMessage_UnableToStartProcess, process.StartInfo.FileName, process.StartInfo.Arguments)));
-                }
+                _disposalTokenSource.SafeCancel();
 
-                await WaitForExitAsync(process, exitedSource.Task, cancellationToken).ConfigureAwait(false);
+                await _completionTask.SafeAwait();
+
+                _disposalTokenSource.Dispose();
             }
-            finally
+
+            public async Task StartAsync(CancellationToken token)
             {
-                process.Exited -= exitedHandler;
+                TaskCompletionSource<object> startCompleteSource = new(TaskCreationOptions.RunContinuationsAsynchronously);
+
+                CancellationToken disposalToken = _disposalTokenSource.Token;
+                _completionTask = Task.Run(() => ExecuteAsync(startCompleteSource, disposalToken), disposalToken);
```

Comment:
Do we not want to link the StartAsync token with the disposal token? In general, we don't really have IsDisposed state checking.

Reply:
If those are linked together and passed into ExecuteAsync, then something could call StartAsync, its Task completes, and then the caller cancels the cancellation token. That would cause the execution of the action to be canceled.

It's my firm believe that the cancellation token that is passed into an async method, unless it is clearly obviously stated and documented otherwise, should only cancel the invocation of the method. Once the method completes, the token that was passed to it should not have any other affect on what that method did.